### ISSUE TITLE
Type Sflong_t is not actually an alias for long long int

### ISSUE THREAD

**Issue by krader1961** [att/ast#457]
If you feel like TL;DR: See my last paragraph.

I've been doing some cleanup work to eliminate lint warnings just because I wanted to do some trivial work that makes the code easier to work with by eliminating obviously bad idioms and bogus warnings. I noticed the following problem because file *src/cmd/ksh93/sh/init.c* contains this line:

```
SH_LINENO->nvalue.llp = &shp->st.lineno;
```

The compiler warns that

```
../src/cmd/ksh93/sh/init.c:1957:27: warning: incompatible pointer types assigning to 'intmax_t *' (aka 'long *') from 'int64_t *' (aka 'long long *') [-Wincompatible-pointer-types]
    SH_LINENO->nvalue.llp = &shp->st.lineno;
                          ^ ~~~~~~~~~~~~~~~
```

This warning is due to file *src/lib/libast/include/sfio.h* containing this definition:

```
/* Sfoff_t should be large enough for largest file address */
#define Sfoff_t intmax_t
#define Sflong_t intmax_t
#define Sfulong_t uintmax_t
#define Sfdouble_t _ast_fltmax_t
```

Notice the comment. Even at the time this code was written it was a certainty that file offsets were 64 bit quantities on platforms that supported that data type.

Note that *src/cmd/ksh93/include/name.h* says in the definition of `union Value` that

```
Sflong_t *llp;  // for long long arithmetic
```

This is where I'd like to swear. What moron thought that `intmax_t` is an alias for `long long int` (or `off_t`)? And who reviewed the change that introduced that bogosity? And why didn't the people who wrote this code pay attention to the compiler warnings they must have been seeing?

**Comment by krader1961** [comment]
P.S., Why is `union Value` even using `Sflong_t`? Since the `llp` member is obviously meant to be a pointer to a 64 bit quantity it should be using `int64_t` as the data type.

**Comment by DavidMorano** [comment]
@krader1961 
> This warning is due to file src/lib/libast/include/sfio.h containing this definition:

Thanks for fixing this.  I and unknown others have come across this problem w/ Sflong_t in the past during builds.  Yes, every compiler warning in the build should eventually be fixed (if we lived in a nice and reasonable world)!  

Thanks.

**Comment by kernigh** [comment]
> What moron thought that intmax_t is an alias for long long int (or off_t)?... And why didn't the people who wrote this code pay attention to the compiler warnings they must have been seeing?

This is system dependent. In my OpenBSD machine, intmax_t, off_t, and int64_t are all aliases of long long. Any system where intmax_t = int64_t would never give a compiler warning. The author who assigned to intmax_t * from int64_t * might have never seen a compiler warning.

Most machines have 64-bit intmax_t and 64-bit int64_t. There's no rule that intmax_t and int64_t must have the same type. If a machine has 64-bit long and 64-bit long long, and one of intmax_t or int64_t is long, and the other is long long, then intmax_t and int64_t have different types. Therefore the assignment to intmax_t * from int64_t * was wrong.

The fix in commit 38d496a changed Sflong_t from intmax_t to int64_t. It also included this change in src/cmd/ksh93/include/name.h, union Value:

```
-    Sflong_t *llp;  // for long long arithmetic
+    int64_t *llp;  // for long long arithmetic
```

The comment is still wrong: int64_t and long long might be different types. It would be wrong to put a long long * in llp, but I found no such wrong code. There is code putting Sflong_t * in llp (in sh/arith.c and sh/name.c), so it assumes that Sflong_t and int64_t are the same type. I wonder if type Sflong_t should die, and the code should just use int64_t.
